github/resource_github_repository_ruleset.go

```diff
@@ -0,0 +1,550 @@
+										Default:     false,
+										Description: "Require an approving review in pull requests that modify files that have a designated code owner. Defaults to `false`.",
+									},
+									"require_last_push_approval": {
+										Type:        schema.TypeBool,
+										Optional:    true,
+										Default:     false,
+										Description: "Whether the most recent reviewable push must be approved by someone other than the person who pushed it. Defaults to `false`.",
+									},
+									"required_approving_review_count": {
+										Type:        schema.TypeInt,
+										Optional:    true,
+										Default:     0,
+										Description: "The number of approving reviews that are required before a pull request can be merged. Defaults to `0`.",
+									},
+									"required_review_thread_resolution": {
+										Type:        schema.TypeBool,
+										Optional:    true,
+										Default:     false,
+										Description: "All conversations on code must be resolved before a pull request can be merged. Defaults to `false`.",
+									},
+								},
+							},
+						},
+						"required_status_checks": {
+							Type:        schema.TypeList,
+							MaxItems:    1,
+							Optional:    true,
+							Description: "Choose which status checks must pass before branches can be merged into a branch that matches this rule. When enabled, commits must first be pushed to another branch, then merged or pushed directly to a branch that matches this rule after status checks have passed.",
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"required_check": {
+										Type:        schema.TypeSet,
+										MinItems:    1,
+										Required:    true,
+										Description: "Status checks that are required. Several can be defined.",
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"context": {
+													Type:        schema.TypeString,
+													Required:    true,
+													Description: "The status check context name that must be present on the commit.",
+												},
+												"integration_id": {
+													Type:        schema.TypeInt,
+													Optional:    true,
+													Description: "The optional integration ID that this status check must originate from.",
+												},
+											},
+										},
+									},
+									"strict_required_status_checks_policy": {
+										Type:        schema.TypeBool,
+										Optional:    true,
+										Description: "Whether pull requests targeting a matching branch must be tested with the latest code. This setting will not take effect unless at least one status check is enabled. Defaults to `false`.",
+									},
+								},
+							},
+						},
+						"non_fast_forward": {
+							Type:        schema.TypeBool,
+							Optional:    true,
+							Description: "Prevent users with push access from force pushing to branches.",
+						},
+						"commit_message_pattern": {
```

Comment:
I guess this is a feature not available trough the UI yet? I couldn't see it when I tested my solution out atleast

Reply:
This is available on org rulesets as well as repo rulesets for repos owned by an org, we should validate  somewhere but it’d have to be outside of the schema’s `validateFunc` I think

Reply:
We could add a sanity check function, lets say `validateRulesetResource()`, which is called as the first step of all the CRUD operations. Something like I did [here](https://github.com/jdamata/terraform-provider-sonarqube/blob/7f9f13f44faccf4a210a09a97729b58bb319c499/sonarqube/resource_sonarqube_portfolio.go#L110-L139)

```go
func validatePortfolioResource(d *schema.ResourceData) error {
        switch selectionMode := d.Get("selection_mode"); selectionMode {
        case "NONE", "MANUAL", "REST":
                return nil


        case "TAGS":
                tags := d.Get("tags").([]interface{})
                if len(tags) == 0 {
                        return fmt.Errorf("validatePortfolioResource: When selection_mode is set to TAGS, you need atleast 1 tag, got: %+v", d.Get("tags"))
                }


                for _, tag := range d.Get("tags").([]interface{}) {
                        tagString := fmt.Sprint(tag)
                        if len(tagString) == 0 {
                                return fmt.Errorf("validatePortfolioResource: When selection_mode is set to TAGS, each tag must be non 0, got: %s", tagString)
                        }
                }
                return nil


        case "REGEXP":
                regexp := d.Get("regexp").(string)
                if len(regexp) == 0 {
                        return fmt.Errorf("validatePortfolioResource: When selection_mode is set to REGEXP, regexp must be set, got: \"%s\"", regexp)
                }
                return nil


        default:
                return fmt.Errorf("resourceSonarqubePortfolioCreate: selection_mode needs to be set to one of NONE, MANUAL, TAGS, REGEXP, REST")
        }
}

func resourceSonarqubePortfolioCreate(d *schema.ResourceData, m interface{}) error {
	if err := checkPortfolioSupport(m.(*ProviderConfiguration)); err != nil {
		return err
	}
       
        // ...
}

```